core/reactor_common.c

```diff
@@ -1316,18 +1316,70 @@ bool _lf_check_deadline(self_base_t* self, bool invoke_deadline_handler) {
     return false;
 }
 
+#ifdef LF_THREADED
+
+void* run_watchdog(watchdog_t* watchdog) {
+    watchdog->thread_active = true;
+    self_base_t* base = watchdog->base;
+
+    while (lf_time_physical() < watchdog->expiration) {
+        interval_t T = watchdog->expiration - lf_time_physical();
+        lf_mutex_unlock(&(base->watchdog_mutex));
```

Comment:
This doesnt seem right. The first iteration of this while loop you are not holding the mutex and can thus not unlock it. Please document how the interaction between this thread running a single watchdog and the worker threads executing reactions should look like.

Reply:
> This doesnt seem right. The first iteration of this while loop you are not holding the mutex and can thus not unlock it. 

Yep, should be fixed now.

> Please document how the interaction between this thread running a single watchdog and the worker threads executing reactions should look like.

Not sure if there is a better way of drawing this out:

- `_lf_invoke_reaction` locks the mutex if the reaction is part of a reactor that contains a watchdog
- `_lf_invoke_reaction` unlocks the mutex after invoking the reaction
- `_lf_watchdog_start` unlocks the mutex only after creating the thread for the watchdog
- `run_watchdog` locks the mutex that was just unlocked by `_lf_watchdog_start`
- `run_watchdog` unlocks->locks the mutex inside the while loop so that other reactions in the reactor can run while the thread is asleep.
- `run_watchdog` unlocks the mutex after executing the watchdog function

Reply:
Would be nice to have it represented as a sequence. In what order to you expect this to happen. I think there is some flaw in your logic here. The watchdog mutex should be held whenever a reaction within that reactor is executed. I don't think it makes sense to unlock that mutex when you call `_lf_watchdog_start` from inside a reaction. Remember that a mutex has to unlocked by the same thread that locks it.

Reply:
Sorry for the delay. My thought process was the following:
Thread 1: If reaction is in a reactor that contains a watchdog, lock mutex - `lf_invoke_reaction`.
Thread 1: Execute reaction.
Thread 1: If reaction body calls `_lf_watchdog_start` , start thread 2.
Thread 1: Inside `_lf_watchdog_start`, unlock mutex.
Thread 2: acquire mutex, set expiration, and unlock mutex
Thread 2: sleep
Thread 1: Inside `_lf_watchdog_start`- if the mutex has been unlocked, lock mutex
Thread 1: Unlock mutex inside `lf_invoke_reaction`.
Thread 2: Once done sleeping, lock mutex, and either reset expiration or execute watchdog function.

In hindsight, I'm now realizing that I need to reset the sleep each time, or I will get undesired behavior past the first watchdog function call.